src/rust/engine/src/externs/interface.rs

```diff
@@ -116,6 +116,11 @@ py_module_initializer!(native_engine, |py, m| {
     "write_stderr",
     py_fn!(py, write_stderr(a: PySession, b: String)),
   )?;
+  m.add(
+    py,
+    "teardown_dynamic_ui",
```

Comment:
I don't think that this method actually needs to be exposed, because the boolean `teardown_ui` parameter could be added to the native `write_stdout` and `write_stderr` methods directly. But you've iterated on this a lot, so won't block.